# 生物信息学实战

## [转录组入门](https://xiapengfei1996.github.io/bioinfo/transcriptome)

[任务清单](https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&mid=2247484843&idx=1&sn=95370393e9d16e09e854adce78c94523&chksm=9b484510ac3fcc064b1bfc19f5cc73a4fbc113a840a40c3a95632127c54f570d9d66550ebc22&scene=21#wechat_redirect)  [参考流程](https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&mid=2247485440&idx=1&sn=392c8649e334af99ac60e903aaec01e9&chksm=9b4848bbac3fc1ad02ac2f11fd9f4440fb13cffc063d19da56fe34317d0621a4972ec26a41d1&mpshare=1&scene=23&srcid=04136Qg8PRlWmG5bwOcYtjwq%23rd)

### 计算机资源的准备

- 系统环境及硬件配置： 
  - macOS High Sierra 10.13.4
  - MacBool Air (13-inch, Early 2015)
  - 处理器   1.6GHz intel Core i5
  - 内存  4GB 1600MHz DDR3
  - 存储  128G
  - `uname -a`  x86_64

- 安装Homebrew包管理，要求预先安装Xcode；
（App Store中Xcode要求macOS High Sierra 10.13.2以上）
```Shell
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

- 用homebrew安装wget；（也可在[rudix网站](rudix.org/packages/wget.html)上下载wget安装包）
```Shell
brew install wget
```

- [wget下载miniconda](https://zhuanlan.zhihu.com/p/25085567)；
此处注意网上教程多在Linux或Windows虚拟Ubantu系统下操作，要注意下载MacOSX版文件！
```Shell
wget https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
```

- bash安装miniconda；
```Shell
bash Miniconda3-latest-MacOSX-x86_64.sh
```

- 输入yes接受条款；

- 回车确认安装位置；

- 输入yes确定PATH；

- source完成最后安装；
此处注意MacOSX系统中不是`.bashrc`，应根据反馈结果修改为`.bash_profile`！
```Shell
source ~/.bash_profile
```

- 查看是否安装成功；
```Shell
conda
```

- 更新miniconda；
```Shell
conda update conda
```

- 添加channels,即为conda增加软件源；
```Shell
conda config - add channels r
conda config - add channels bioconda
conda config - add channels defaults
conda config – add channels conda-forge
```

- 查看添加的频道；
```Shell
conda config – get channels
```

- 添加清华源镜像；
```Shell
conda config - add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
conda config - add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
conda config - add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/msys2/
```

- 显示已添加的源镜像地址；
```Shell
conda config --set show_channel_urls yes
```

- 安装sratoolkit,fastqc,hisats,samtools,htseq-count,R,Rstudio软件；
```Shell
conda install -c bioconda sra-tools
conda install -c bioconda fastqc
conda install -c bioconda hisat2
conda install -c bioconda samtools
conda install -c bioconda htseq
# 本机已安装过R和RStudio,故未重复下载
# conda install r
# conda install rstudio
```

### 读文章拿到测序数据

- 检索实战学习示例文章；
[AKAP95 regulates splicing through scaffolding RNAs and RNA processing factors. Nat Commun 2016 Nov 8;7:13347. PMID: 27824034](https://www.nature.com/articles/ncomms13347)

- 阅读文章Methods部分，尤Bioinformatic analyses和Data availability部分；
> Bioinformatic analyses
> We obtained 35–55 million of paired 51 bp reads for each sample. The investigator was blinded to sample identities when performing bioinformatic analyses. All the reads were mapped to the **human reference genome (GRCh37/hg19)** using **TopHat (v2.0.13)** with a **gene transfer file (GTF version GRCh37.70)**. **Low quality mapped reads (MQ>30)** were removed from the analysis. The **mean insert sizes and the s.d.’s** were calculated using [**Picard-tools (v1.126)**](http://broadinstitute.github.io/picard/).
> For mRNA-Seq, **read count tables** were generated using **HTSeq (v0.6.0)53** and **deferential expression (DE) analysis of genes** were performed using **DESeq (v3.0)54**. **Deferential exon usage** was performed using **DEXSeq (v3.1)29**. Two independent biological repeats for control and KD were included in both DEseq and DEXseq. For 293 cells, the mRNA-seq results of the control samples include (1) those from the doxycycline-treated parental Flp-In T-REx 293 cells by us and (2) those from the doxycycline-treated control Flp-In T-REx 293 cells performed by another group unrelated to us (sample GSM1095127 in GSE44976)31, and the two AKAP95 KD samples were from the two different clones (miR#8 and miR#12) of Flp-In T-REx 293 cells stably expressing AKAP95 microRNAs after doxycycline treatment. For mouse ES cells, the two biological repeats for both control and shRNA#1-mediated Akap95 KD were independently generated by two different persons in our lab. The **read per million normalized BigWig files** were generated using **BEDTools (v2.17.0)55** and **bedGraphToBigWig tool (v4)**. All the **downstream statistical analyses** and generating **plots** were performed in [**R (v3.1.1)**](http://www.r-project.org/). **Gene ontology analysis** was performed at [**DAVID**](http://david.ncifcrf.gov/).
> For RIP-Seq, mapped files were passed to Model-based Analysis of ChIP-Seq (MACS) (v1.4.2 20120305)56 for peak calling and [HOMER (vv3.12, 6-8-2012)](http://homer.salk.edu/homer/)57 was used for Motif finding. Bedgraph files generated by MACS were then normalized based on read per million and converted to BigWig files using bedGraphToBigWig. The input RNA-seq files were not subject to MACS. The genomic regions for annotation (intergenic regions, exons, UTRs, distal and proximal introns) were extracted from the GTF file used above. Peak annotations and coverage calculations for these genomic regions were performed using Bedtools and SAMtools (v0.1.18)58, respectively. Profile plots were generated by ngs.plot (v2.47)59.

> Data availability
> The RIP-seq an RNA-seq data have been deposited in the Gene Expression Omnibus database, with accession code **GSE81916**. All other data is available from the author upon reasonable request.

- 在[GEO](https://www.ncbi.nlm.nih.gov/geo/)上查找[**GSE81916**](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81916),查找得其对应的SRA号为**SRP075747**（以及BioProject号**PRJNA323422**）；

> Overall design
> Samples 1-8 are RNA-immunoprecipitation (RIP)-seq to determine AKAP95 binding to the transcriptome. **Samples 9-15 are mRNA-seq** to determine effect of AKAP95 knockdown in human 293 cells (9-11) or mouse ES cells (12-15).

- 阅读网页Overall design部分介绍得知，属于RNA-seq数据的是第9-15个样品；

- 在[SRA](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=search_obj)上查找[**SRP075747**](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP075747&go=go)（或[**PRJNA323422**](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA323422)）,查找得其第9-15个样品，即**GSM2177723**到**GSM2177729**对应的SRA号为**SRR3589956**到**SRR3589962**；
_注意理解GEO/SRA数据库的数据存放形式_

- 安装axel来下载数据比wget及prefetch更快；（也可使用[GEOquery](http://www.bio-info-trainee.com/bioconductor_China/software/GEOquery.html)下载）
（Aspera只有Windows系统下的client）
```Shell
brew install axel
```

- 循环axel语句实现数据下载；
（**2200+KB/s，用时约两小时**）
```Shell
for i in {56..62}
do
axel ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP075/SRP075747/SRR35899$i/SRR35899$i.sra
done
```

### 了解fastq测序数据

- 循环sratoolkit（sra-tools）中fastq-dump语句把sra文件转换为fastq格式；
（**用时约三个半小时**）
此处注意fastq-dump命令不显示百分数进度条！
```Shell
fastq-dump -h # 查看帮助文件

for i in {56..62}
do
fastq-dump --gzip --split-3 -O /Users/xiapengfei/ -A SRR35899${i}.sra
done
```

- 循环fastqc语句测试测序文件的质量；([MultiQC可以批量显示质控结果](http://fbb84b26.wiz03.com/share/s/3XK4IC0cm4CL22pU-r1HPcQQ1iRTvV2GwkwL2AaxYi2fXHP7)）
（**4线程，用时约25分钟**）
```Shell
for i in {56..62}
do
fastqc -t 4 -O /Users/xiapengfei/ SRR35899$i.sra_1.fastq.gz SRR35899$i.sra_2.fastq.gz
done
```
_[理解测序reads，GC含量，质量值，接头，index，fastqc的全部报告](https://github.com/XIAPengfei1996/Shia/tree/master/bioinfo/fastqc_html)_

[**经验教训**](http://www.bio-info-trainee.com/2275.html)

### 了解参考基因组及基因注释

- 在[UCSC](http://genome.ucsc.edu/index.html)中打开Downloads中Genome Data页面，点击Human选择Feb. 2009 (GRCh37/hg19)栏下Full data set，下拉找到chromFa.tar.gz文件；

> chromFa.tar.gz - The assembly sequence in one file per chromosome.
>     Repeats from RepeatMasker and Tandem Repeats Finder (with period
>     of 12 or less) are shown in lower case; non-repeating sequence is
>     shown in upper case.

- axel命令下载hg19参考基因组数据；
（**400+KB/s，用时约35分钟**）
```Shell
axel http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz
tar zxvf chromFa.tar.gz  # 解压该文件
cat *.fa > hg19.fa  # 将解压文件合并为一个文件
rm chr*.fa  # 删除无关数据
```

- 从[gencode数据库](http://www.gencodegenes.org)中打开Data中Human栏内[GRCh37-mapped Releases](https://www.gencodegenes.org/releases/grch37_mapped_releases.html)，点击GENCODE release栏下最新版的数据，下载基因注释文件；
（**用时5分钟**)
```Shell
axel ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/GRCh37_mapping/gencode.v28lift37.annotation.gtf.gz  # GTF格式主要是用来描述基因的注释
axel ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/GRCh37_mapping/gencode.v28lift37.annotation.gff3.gz  # GFF文件是一种用来描述基因组特征的文件，现在我们所使用的大部分都是第三版（GFF3）
# gzip -d解压并删除压缩包
gzip -d gencode.v28lift37.annotation.gtf.gz
gzip -d gencode.v28lift37.annotation.gff3.gz
```

_- 用IGV去查看感兴趣的基因的结构；_
  - 下载安装[JAVA 8](https://java.com/en/download/mac_download.jsp)；
  - 下载[igv](http://software.broadinstitute.org/software/igv/download)直接打开；
  - 在菜单栏Genomes中选择Load Genome from File…导入hg19.fa文件；
  - 在菜单栏Tools中打开Run igvtools…界面，在Command栏中选择Sort，在Input File栏中导入gtf文件后运行；
  - 在菜单栏File中选择Load from File…导入sorted.gtf文件并index；
  - 在工具栏输入查找的基因名词（如：TP53）或染色体定位（如chr17:7,569,720-7,592,868)查看可视化结构；

### 序列比对

从[HISAT2](http://ccb.jhu.edu/software/hisat2/index.shtml)的主页下载index文件
```Shell
axel ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/hg19.tar.gz
```
把fastq格式的reads比对上去得到sam文件

用samtools把sam文件转为bam文件，并且排序(注意N和P两种排序区别)索引好

### reads计数


## [ChIP-seq基础入门](https://xiapengfei1996.github.io/bioinfo/ChIP-seq)

## [450K甲基化芯片数据处理](https://xiapengfei1996.github.io/bioinfo/Methylation450K_BeadChip)

## [lncRNA数据分析](https://xiapengfei1996.github.io/bioinfo/lncRNA)
