# 生物信息学实战

## [转录组入门](https://xiapengfei1996.github.io/bioinfo/transcriptome)

### 计算机资源的准备

- 系统环境及硬件配置： 
  - macOS High Sierra 10.13.4
  - MacBool Air (13-inch, Early 2015)
  - 处理器   1.6GHz intel Core i5
  - 内存  4GB 1600MHz DDR3
  - 存储  128G
  - `uname -a`  x86_64

- 安装Homebrew包管理，要求预先安装Xcode；
（App Store中Xcode要求macOS High Sierra 10.13.2以上）
```Shell
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

- 用homebrew安装wget；（也可在[rudix网站](rudix.org/packages/wget.html)上下载wget安装包）
```Shell
brew install wget
```

- [wget下载miniconda](https://zhuanlan.zhihu.com/p/25085567)；
此处注意网上教程多在Linux或Windows虚拟Ubantu系统下操作，要注意下载MacOSX版文件！
```Shell
wget https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
```

- bash安装miniconda；
```Shell
bash Miniconda3-latest-MacOSX-x86_64.sh
```

- 输入yes接受条款；

- 回车确认安装位置；

- 输入yes确定PATH；

- source完成最后安装；
此处注意MacOSX系统中不是`.bashrc`，应根据反馈结果修改为`.bash_profile`！
```Shell
source ~/.bash_profile
```

- 查看是否安装成功；
```Shell
conda
```

- 更新miniconda；
```Shell
conda update conda
```

- 添加channels,即为conda增加软件源；
```Shell
conda config - add channels r
conda config - add channels bioconda
conda config - add channels defaults
conda config – add channels conda-forge
```

- 查看添加的频道；
```Shell
conda config – get channels
```

- 添加清华源镜像；
```Shell
conda config - add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
conda config - add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
conda config - add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/msys2/
```

- 显示已添加的源镜像地址；
```Shell
conda config --set show_channel_urls yes
```

- 安装sratoolkit,fastqc,hisats,samtools,htseq-count,R,Rstudio软件；
```Shell
conda install -c bioconda sra-tools
conda install -c bioconda fastqc
conda install -c bioconda hisat2
conda install -c bioconda samtools
conda install -c bioconda htseq
# 本机已安装过R和RStudio,故未重复下载
# conda install r
# conda install rstudio
```

### 读文章拿到测序数据

- 检索实战学习示例文章；
[AKAP95 regulates splicing through scaffolding RNAs and RNA processing factors. Nat Commun 2016 Nov 8;7:13347. PMID: 27824034](https://www.nature.com/articles/ncomms13347)

- 阅读文章Methods部分，尤Bioinformatic analyses和Data availability部分；
> Bioinformatic analyses
> We obtained 35–55 million of paired 51 bp reads for each sample. The investigator was blinded to sample identities when performing bioinformatic analyses. All the reads were mapped to the **human reference genome (GRCh37/hg19)** using **TopHat (v2.0.13)** with a **gene transfer file (GTF version GRCh37.70)**. **Low quality mapped reads (MQ>30)** were removed from the analysis. The **mean insert sizes and the s.d.’s** were calculated using [**Picard-tools (v1.126)**](http://broadinstitute.github.io/picard/).
> For mRNA-Seq, **read count tables** were generated using **HTSeq (v0.6.0)53** and **deferential expression (DE) analysis of genes** were performed using **DESeq (v3.0)54**. **Deferential exon usage** was performed using **DEXSeq (v3.1)29**. Two independent biological repeats for control and KD were included in both DEseq and DEXseq. For 293 cells, the mRNA-seq results of the control samples include (1) those from the doxycycline-treated parental Flp-In T-REx 293 cells by us and (2) those from the doxycycline-treated control Flp-In T-REx 293 cells performed by another group unrelated to us (sample GSM1095127 in GSE44976)31, and the two AKAP95 KD samples were from the two different clones (miR#8 and miR#12) of Flp-In T-REx 293 cells stably expressing AKAP95 microRNAs after doxycycline treatment. For mouse ES cells, the two biological repeats for both control and shRNA#1-mediated Akap95 KD were independently generated by two different persons in our lab. The **read per million normalized BigWig files** were generated using **BEDTools (v2.17.0)55** and **bedGraphToBigWig tool (v4)**. All the **downstream statistical analyses** and generating **plots** were performed in [**R (v3.1.1)**](http://www.r-project.org/). **Gene ontology analysis** was performed at [**DAVID**](http://david.ncifcrf.gov/).
> For RIP-Seq, mapped files were passed to Model-based Analysis of ChIP-Seq (MACS) (v1.4.2 20120305)56 for peak calling and [HOMER (vv3.12, 6-8-2012)](http://homer.salk.edu/homer/)57 was used for Motif finding. Bedgraph files generated by MACS were then normalized based on read per million and converted to BigWig files using bedGraphToBigWig. The input RNA-seq files were not subject to MACS. The genomic regions for annotation (intergenic regions, exons, UTRs, distal and proximal introns) were extracted from the GTF file used above. Peak annotations and coverage calculations for these genomic regions were performed using Bedtools and SAMtools (v0.1.18)58, respectively. Profile plots were generated by ngs.plot (v2.47)59.

> Data availability
> The RIP-seq an RNA-seq data have been deposited in the Gene Expression Omnibus database, with accession code **GSE81916**. All other data is available from the author upon reasonable request.

- 在[GEO](https://www.ncbi.nlm.nih.gov/geo/)上查找[**GSE81916**](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81916),查找得其对应的SRA号为**SRP075747**（以及BioProject号**PRJNA323422**）；

> Overall design
> Samples 1-8 are RNA-immunoprecipitation (RIP)-seq to determine AKAP95 binding to the transcriptome. **Samples 9-15 are mRNA-seq** to determine effect of AKAP95 knockdown in human 293 cells (9-11) or mouse ES cells (12-15).

- 阅读网页Overall design部分介绍得知，属于RNA-seq数据的是第9-15个样品；

- 在[SRA](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=search_obj)上查找[**SRP075747**](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP075747&go=go)（或[**PRJNA323422**](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA323422)）,查找得其第9-15个样品，即**GSM2177723**到**GSM2177729**对应的SRA号为**SRR3589956**到**SRR3589962**；
_注意理解GEO/SRA数据库的数据存放形式_

- 安装axel来下载数据比wget及prefetch更快；（也可使用[GEOquery](http://www.bio-info-trainee.com/bioconductor_China/software/GEOquery.html)下载）
（Aspera只有Windows系统下的client）
```Shell
brew install axel
```

- 循环语句实现数据下载
（**用时约两小时**）
```Shell
for i in {56..62}
do
axel ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP075/SRP075747/SRR35899$i/SRR35899$i.sra
done
```

### 了解fastq测序数据

- 循环sratoolkit（sra-tools）中fastq-dump语句把sra文件转换为fastq格式
（**用时约三个半小时**）
此处注意fastq-dump命令不显示百分数进度条！
```Shell
# 查看帮助文件
fastq-dump -h

for i in {56..62}
do
fastq-dump --gzip --split-3 -O /Users/xiapengfei/ -A SRR35899${i}.sra
done
```

3：26
- 循环fastqc语句测试测序文件的质量
（**用时约25分钟**）
```Shell
for i in {56..62}
do
fastqc -t 4 -O /Users/xiapengfei/ SRR35899$i.sra_1.fastq.gz SRR35899$i.sra_2.fastq.gz
done
```
_理解测序reads，GC含量，质量值，接头，index，fastqc的全部报告_



## [ChIP-seq基础入门](https://xiapengfei1996.github.io/bioinfo/ChIP-seq)

## [450K甲基化芯片数据处理](https://xiapengfei1996.github.io/bioinfo/Methylation450K_BeadChip)

## [lncRNA数据分析](https://xiapengfei1996.github.io/bioinfo/lncRNA)
